### ISSUE TITLE
Adding scalajs-linker as a dependency introduces unexpected error

### ISSUE THREAD

**Issue by kincsescsaba** [scala/scala3#21560]
I ran the Scala 3 compiler (version: 3.5.0) successfully with the `-scalajs` option, `sjsir` files appear in the target directory as expected.

After this step, I wanted to create human-readable JavaScript. Therefore I added `scalajs-linker` as a dependency to the SBT project.

Running the first step again with the `scalajs-linker` being added, I unexpectedly get an error:

```
     while compiling: (inline)
        during phase: genSJSIR
                mode: Mode(ImplicitsEnabled)
     library version: version 2.13.14
    compiler version: version 3.5.0
            settings: -classpath "" -d (memory) -nowarn true -scalajs true -usejavacp true

An exception or error caused a run to abort: 'org.scalajs.ir.Types$ClassType org.scalajs.ir.Types$ClassType$.unapply(org.scalajs.ir.Types$ClassType)' 
java.lang.NoSuchMethodError: 'org.scalajs.ir.Types$ClassType org.scalajs.ir.Types$ClassType$.unapply(org.scalajs.ir.Types$ClassType)'
...
```

I assume the source to be compiled was OK, as `sjsir` files were generated.

Dependencies were set up with cross-version settings in SBT:
```
libraryDependencies ++= Seq(
      "org.scala-lang" %% "scala3-compiler" % scalaVersion.value,
      ("org.scala-js" %% "scalajs-library" % "1.16.0").withCrossVersion(CrossVersion.for3Use2_13),
      ("org.scala-js" %% "scalajs-linker" % "1.16.0").withCrossVersion(CrossVersion.for3Use2_13),
       ...
    )

```
The `scalajs-library` dependency did not introduce binary compatibility issues (Scala 2.13 and 3 should be binary compatible  [https://docs.scala-lang.org/overviews/core/binary-compatibility-of-scala-releases.html](https://docs.scala-lang.org/overviews/core/binary-compatibility-of-scala-releases.html)) while the linker (or an IR related dependency of it) did.

Looks like a nested case does not get an `unapply` method generated when it was compiled, this may cause the linker jar file to be binary incompatible, if there is no other circumstance that could cause this error.

**Comment by sjrd** [comment]
I believe you're mixing two unrelated things. One does not add the linker to a project to produce JavaScript. Instead we add the `sbt-scalajs` plugin as an sbt plugin, then add `enablePlugins(ScalaJSPlugin)` to the `project` that should be a JS project. In fact we don't even manually add `-scalajs`. The sbt plugin takes care of that. The linker must not appear in the dependencies of your project.

I suggest you follow a Scala.js tutorial or template project for the correct setup.

**Comment by kincsescsaba** [comment]
Thanks for looking into this. 

I use a customized approach extending `dotty.tools.dotc.Driver` as I need on-the-fly arbitrary binary and JS generation. The mentioned standard solution, which I know of, seems to be for a fixed, non-changing codebase, which shall be in a JS project, as you have mentioned, so it's not the right fit for what I need.

I assumed the codebase to be self-documenting, and what can be found suggested to go in that direction.

If the plugin works with the latest versions, respectively 3.5.0 for Scala compiler and 1.16.0 for ScalaJs, then I could find the needed parts in plugin internals.

I'll check and revert if there is more information to see the reasons for this crash.

**Comment by sjrd** [comment]
OK, then it's because you mix two incompatible versions of the `scalajs-ir` artifact's contents: one through the linker and one through the compiler. Either downgrade `scalajs-linker` to the same version that was used to build the compiler you're using. Or isolate the linker from the compiler by putting them on separate projects/classpaths.

Note the `scalajs-ir` artifact can break binary compatibility between minor releases, according to our versioning policy: 
https://github.com/scala-js/scala-js/blob/v1.16.0/VERSIONING.md

**Comment by kincsescsaba** [comment]
I'll do that. After further examining this issue, I found that we're not dealing with a missing `unapply` method but one generated with a different return type which I'll need to check if it is a bug or not. I'll check the details, as that could be a more general problem while closing this ticket and opening a new one if needed, as we've discussed the parts that were related to the ScalaJs-related side of this.
